assertk/src/jvmMain/kotlin/assertk/assertions/any.kt

```diff
@@ -100,8 +121,10 @@ fun <T : Any> Assert<T>.isDataClassEqualTo(expected: T) = given { actual ->
 private fun <T> Assert<T>.isDataClassEqualToImpl(expected: T, kclass: KClass<*>?): Unit = given { actual ->
     if (actual == expected) return
     if (kclass != null && kclass.isData) {
-        for (prop in kclass.memberProperties) {
-            prop(prop).isDataClassEqualToImpl(prop.call(expected), prop.returnType.classifier as? KClass<*>)
+        for (memberProp in kclass.memberProperties) {
+            @Suppress("UNCHECKED_CAST")
+            val force = memberProp as KProperty1<T, Any?>
```

Comment:
Does this cast always succeed? I wonder why `kclass.memberProperties` returns `KCallable`.

Reply:
This one is rather odd.
The `klass.memberProperties` returns `Collection<KProperty1<T, *>>`, which seems to be fine. But when you try to get an item from this collection, like here in the for-loop, the type of the item is inferred to `KProperty1<out Any, *>`. And this type does not suit the signature of `prop(KProperty1)` at all. That's why there is a cast.

I assume that the cast must always succeed for several reasons. First of all, it is the same cast that was already used in another more general function here: https://github.com/willowtreeapps/assertk/blob/master/assertk/src/jvmMain/kotlin/assertk/assertions/any.kt#L125
Secondly, the generic parameter of the `kclass` always corresponds to the `T`, even though it is star-projected. This is because we start with the top-most data-class type in the user-facing `isDataClassEqualTo`. Then, whenever we go down into one of the properties, we use the property return type to keep the correspondence.
Last, but not least, the tests for the `isDataClassEqualTo`, which include nested data classes, still pass.